File: src/nl/wernerkroneman/Drawy/DrawyMcDrawFace.java
package nl.wernerkroneman.Drawy;/*
 * main.cpp
 *
 *  Created on: 10 jun. 2016
 *      Author: werner
 */

import com.jogamp.opengl.GLCapabilities;
import com.jogamp.opengl.GLProfile;
import com.jogamp.opengl.awt.GLCanvas;
import nl.wernerkroneman.Drawy.AbstractToConcreteConverter.AbstractToConcrete;
import nl.wernerkroneman.Drawy.ConcreteModelling.AABB;
import nl.wernerkroneman.Drawy.ConcreteModelling.Scene;
import nl.wernerkroneman.Drawy.GlRenderer.GlVisualiser;
import nl.wernerkroneman.Drawy.ModelEditor.BlockingInteractorInterface;
import nl.wernerkroneman.Drawy.ModelEditor.DescriptionSession;
import nl.wernerkroneman.Drawy.ModelEditor.DescriptionSessionListener;
import nl.wernerkroneman.Drawy.ModelEditor.Knowledge;
import nl.wernerkroneman.Drawy.Modelling.Model;
import org.joml.Vector3d;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

public class DrawyMcDrawFace {

    public static void main(String[] args) {
        JFrame frame = new JFrame();

        frame.setTitle("DrawyMcDrawface");

        GLProfile glprofile = GLProfile.getGL2GL3();
        GLCapabilities glcapabilities = new GLCapabilities(glprofile);
        final GLCanvas glcanvas = new GLCanvas(glcapabilities);

        frame.getContentPane().add(glcanvas);

        frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);

        frame.setSize(800, 600);

        UserInteractor interactor = new UserInteractor();
        frame.getContentPane().add(interactor, BorderLayout.SOUTH);

        frame.setVisible(true);

        // Initialize empty knowledge
        Knowledge knowledge = Knowledge.knowledgeWithPrimitives();

        System.out.println("Loaded " + knowledge.getNumberOfObjects() + " concepts. ");

        BlockingInteractorInterface iface = new BlockingInteractorInterface(interactor);
        DescriptionSession session = DescriptionSession.createDescriptionSession(knowledge, iface);

        // Create a visualizer
        GlVisualiser visualiser = new GlVisualiser();

        glcanvas.addGLEventListener(visualiser);

        // Create a converter that uses meshes compatible with the Renderer
        AbstractToConcrete converter = new AbstractToConcrete(visualiser.getMeshFactory());

        // Every time changes are made to the scene
        session.addListener(new MyDescriptionSessionListener(converter, visualiser, glcanvas));

        System.out.print("Main scene" + ">: ");

        Timer timer = new Timer(50, new ActionListener() {
            double t = 0;

            @Override
            public void actionPerformed(ActionEvent e) {
                t += 0.01;
                AABB box = visualiser.getScene().getRootSceneNode().computeWorldAABB();
                double d = 10 * Math.max(box.getWidth(), box.getDepth());
                Scene.EYE.set(d * Math.sin(t), 10, d * Math.cos(t));
                Scene.LOOKAT_CENTER.set(box.getCenter(new Vector3d()));
                glcanvas.display();
            }
        });
        timer.start();

        session.start();
    }

    private static class MyDescriptionSessionListener implements DescriptionSessionListener {
        private final AbstractToConcrete converter;
        private final GlVisualiser visualiser;
        private final GLCanvas glcanvas;

        public MyDescriptionSessionListener(AbstractToConcrete converter, GlVisualiser visualiser, GLCanvas glcanvas) {
            this.converter = converter;
            this.visualiser = visualiser;
            this.glcanvas = glcanvas;
        }

        @Override
        public void modelChanged(Model scene) {
            // Convert it to a concrete scene
            Scene concreteScene = converter.computeScene(scene);

            // Pass it to
            visualiser.setScene(concreteScene);

            glcanvas.display();
        }

        @Override
        public void sessionEnded() {

        }

        @Override
        public void recursiveSessionStarted(DescriptionSession recursiveSession) {

        }
    }
}

File: src/main/java/nl/wernerkroneman/Drawy/DrawyMcDrawFace.kt
/*
 * Copyright (c) 2017 Werner Kroneman
 *
 * This file is part of DrawyMcDrawface.
 *
 * DrawyMcDrawface is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DrawyMcDrawface is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with DrawyMcDrawface.  If not, see <http://www.gnu.org/licenses/>.
 */

package nl.wernerkroneman.Drawy

/*
 * main.cpp
 *
 *  Created on: 10 jun. 2016
 *      Author: werner
 */

import com.jogamp.opengl.GLCapabilities
import com.jogamp.opengl.GLProfile
import com.jogamp.opengl.awt.GLCanvas
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.Single.just
import nl.wernerkroneman.Drawy.AbstractToConcreteConverter.AbstractToConcrete
import nl.wernerkroneman.Drawy.ConcreteModelling.Scene
import nl.wernerkroneman.Drawy.GlRenderer.GlVisualiser
import nl.wernerkroneman.Drawy.Interface.BlockingInteractor
import nl.wernerkroneman.Drawy.Interface.UserInteractor
import nl.wernerkroneman.Drawy.Modelling.ModelSpecification
import java.awt.BorderLayout
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.util.concurrent.TimeUnit
import javax.swing.JFrame
import javax.swing.Timer
import javax.swing.WindowConstants

fun main(args: Array<String>) {

    val frame = JFrame("DrawyMcDrawface")

    val glprofile = GLProfile.getGL2GL3()
    val glcapabilities = GLCapabilities(glprofile)
    val glcanvas = GLCanvas(glcapabilities)

    frame.contentPane.add(glcanvas)

    frame.defaultCloseOperation = WindowConstants.EXIT_ON_CLOSE

    frame.setSize(800, 600)

    val interactor = UserInteractor()
    frame.contentPane.add(interactor, BorderLayout.SOUTH)

    frame.isVisible = true

    frame.pack()

    val utilityPole = ModelSpecification()

    val abstractScene = just(utilityPole)

    val timer = Observable.interval(1000 / 60, TimeUnit.MILLISECONDS)

    val iface = BlockingInteractor(interactor)
    //val session = DescriptionSession(
    //       interpreter = defaultModelInterpreter,
    //       interactorIface = iface)

    // Create a visualizer
    val visualiser = GlVisualiser()

    glcanvas.addGLEventListener(visualiser)


    // Every time changes are made to the scene
    abstractScene
            .map({ Scene(AbstractToConcrete(visualiser.meshFactory).concreteForModel(it)) })
            .subscribe { scene -> visualiser.scene = scene; glcanvas.display() }


    /*internal var t = 0.0

    override fun actionPerformed(e: ActionEvent) {
        t += 0.05
        val box = visualiser.scene.rootSceneNode.aabb
        val d = 5 * Math.max(box.sizeX, box.sizeZ)
        Scene.EYE.set(d * Math.sin(t), 5.0 * Math.sin(t), d * Math.cos(t))
        Scene.LOOKAT_CENTER.set(box.center)
        glcanvas.display()
    }*/
}

